feat(CanonicalEnsemble):add params to CanonicalEnsemble struct + refactor

or4nge19 opened the pull request:
This PR refines the general measure-theoretic framework for the canonical ensemble, applicable to both classical continuous systems (like an ideal gas) and discrete systems (like a spin lattice). It introduces additional parameters to the CanonicalEnsemble structure and various distinctions between entropy and proves some key thermodynamics identities.

We follow the *semi-classical* approach in Landau-Lefschetz and also in Tong's lecture notes in designing a general API (for this initial level of analysis) to handle both discrete and continuous systems. The explicitation of a classical principled exposition like L&L constrained us to distinguish two layers:

Mathematical layer: Raw partition function `∫ exp(-βE) dμ`, probability densities w.r.t. base measure  
Physical layer: Dimensionless thermodynamic quantities via `phaseSpaceUnit^dof` normalization

This seems to be a more clear way to analyze the presuppositions of the classical theory that pre-empt ambiguity issues (Gibbs paradox) by making entropy absolute rather than relative.

 Main additions :

New parameters to the `CanonicalEnsemble` structure:
- `phaseSpaceUnit : ℝ` (physically Planck's constant h)  
- `dof : ℕ` (degrees of freedom)

Thermodynamic quantities:
- `partitionFunction = mathematicalPartitionFunction / h^dof`
- `helmholtzFreeEnergy = -kT log Z` 
- `thermodynamicEntropy` (absolute) vs `differentialEntropy` (can be negative)
- `physicalProbability = probability * h^dof` (dimensionless)

 Main results and corollaries from L&L and Tong's notes that we prove in  the file`Lemmas`  are:

-  *Fundamental identity*: `F = U - TS` (theorem `helmholtzFreeEnergy_eq_meanEnergy_sub_temp_mul_thermodynamicEntropy`)  
-  *Response formula**: `U = -∂log Z/∂β` (theorem `meanEnergy_eq_neg_deriv_log_mathZ_of_beta`)  
-  *Entropy relations*: `S_thermo = S_diff - k·dof·log h` with semi-classical correction

 Finite case (`IsFinite`) specializations

Specialization where `dof = 0`, `phaseSpaceUnit = 1`, where we show:
- Shannon entropy = differential entropy = thermodynamic entropy
- No semi-classical corrections needed
- Unconditional entropy non-negativity

As designed here the introduction of this necessary semi-classical correction `k·dof·log h` appears systematically where needed and vanishes in the discrete case.

or4nge19 committed:
feat(CanonicalEnsemble):add phase_space_unit and dof parameters and refactor

or4nge19 committed:
Update Finite.lean

refactor

or4nge19 committed:
Merge pull request #708 from HEPLean/master

merge master

or4nge19 committed:
Merge pull request #707 from HEPLean/or4nge19-HelmotzFreeEnergy

feat(CanonicalEnsemble):add phase_space_unit and dof parameters and r…

or4nge19 committed:
Update Basic.lean

or4nge19 committed:
Update Basic.lean

or4nge19 committed:
re-included some prob lemmas removed by mistake

or4nge19 committed:
re-ncluded some previous doc-stings

or4nge19 committed:
add Helmholtz identity

or4nge19 committed:
add Thermodynamics identities, Helmholtz identity

or4nge19 committed:
fix lint

or4nge19 committed:
refactor TwoState according to new CanonicalEnsemble paramters

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Basic.lean:
```diff
@@ -254,7 +334,7 @@ lemma μBolt_nsmul [SigmaFinite 𝓒.μ] (n : ℕ) (T : Temperature) :
     exact MeasureTheory.measurePreserving_piFinSuccAbove (fun _ => 𝓒.μBolt T) 0
 
-lemma μBolt_ne_zero_of_μ_ne_zero (T : Temperature) (h : 𝓒.μ ≠ 0) :
+  lemma μBolt_ne_zero_of_μ_ne_zero (T : Temperature) (h : 𝓒.μ ≠ 0) :
```
Can we remove this added space?

jstoobysmith replied:
Still to be done.

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Basic.lean:
```diff
@@ -631,17 +700,309 @@ lemma differentialEntropy_nonneg_of_prob_le_one
+temperature. It is intentionally retained (even though not needed for mere division by `T.val`)
+so that subsequent lemmas may rely on it (e.g. for monotonicity, differentiability,
+or asymptotic considerations near absolute zero). We mark it as an unused argument explicitly. -/
+noncomputable def thermodynamicEntropy (T : Temperature) (hT : 0 < T.val) : ℝ :=
```
Here I still think we should define the entropy as L&L do it in equation 7.12 and then show that it is equal to this.

or4nge19 replied:
yes 100%

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Basic.lean:
```diff
@@ -631,17 +700,309 @@ lemma differentialEntropy_nonneg_of_prob_le_one
+
+/-- An intermediate potential defined from the mathematical partition function. See
+`helmholtzFreeEnergy` for the physical thermodynamic quantity. -/
+noncomputable def helmholtzPotential (T : Temperature) : ℝ :=
```
Wouldn't mathematicalHelmholtzFreeEnergy be more consistent?

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Basic.lean:
```diff
@@ -12,32 +12,84 @@ import Mathlib.Analysis.SpecialFunctions.Log.Summable
+  /-- The number of degrees of freedom, used to make the partition function dimensionless.
+  For a classical system of N particles in 3D, this is `3N`. For a system of N spins,
+  this is typically `0` as the state space is already discrete. -/
+  dof : ℕ
```
It would be good if we could some how get the dof from the dimension of iota, but I understand this might be difficult and what you have done might be the best solution here.

or4nge19 replied:
Sure, though I'm happy to postpone it as it requires building a significant hierarchy of typeclasses to teach Lean what a 'phase space' is. We can put it among the TODOs (there are a few I want to add at the last commit), but I suspect the need will awyway arise soon in the development

jstoobysmith replied:
sounds good to me

or4nge19 committed:
refactor, thermodynamicEntropy

we make entropy's statistical definition primary and the thermodynamic identity a provable consequence
we also add meanEnergy_eq_neg_deriv_log_mathZ_of_beta and other basic lemmas for the defs introduced

or4nge19 committed:
fix lints

or4nge19 committed:
split Basic.lean into Defs + basic - fix lints

or4nge19 committed:
rename files

or4nge19 committed:
Update Lemmas.lean

or4nge19 committed:
fix lints

or4nge19 committed:
fix lints

or4nge19 committed:
fix lints

or4nge19 committed:
fix lints

or4nge19 committed:
fix lints + update module doc strings

or4nge19 committed:
fix lints

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Basic.lean:
```diff
@@ -10,34 +10,120 @@ import PhysLean.Meta.Informal.SemiFormal
+  This constant is necessary to define an absolute (rather than relative) thermodynamic
+  entropy. In the semi-classical approach, this unit is identified with Planck's constant `h`.
+  For discrete systems with a counting measure, this unit should be set to `1`. -/
+  phase_space_unit : ℝ := 1
```
Adding a comment, so we remember to rename

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Basic.lean:
```diff
@@ -631,17 +733,300 @@ lemma differentialEntropy_nonneg_of_prob_le_one
+/-- The dimensionless physical probability density. This is obtained by dividing the
+phase space measure by the fundamental unit `h^dof`, making the probability
+density `ρ_phys = ρ_math * h^dof` dimensionless. -/
+noncomputable def physicalProbability (T : Temperature) (i : ι) : ℝ :=
```
In the doc-string can we say it is the probability density with respect to the measure....

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Lemmas.lean:
```diff
@@ -0,0 +1,735 @@
+open scoped Topology ENNReal
+
+/-- Positivity of `β` from positivity of temperature. -/
+lemma beta_pos (T : Temperature) (hT_pos : 0 < T.val) : 0 < (T.β : ℝ) := by
```
Could we move this to the Temperature file

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Lemmas.lean:
```diff
@@ -0,0 +1,735 @@
+see: Tong (§1.3.2, §1.3.3), L&L (§31, implicitly, and §36)
+Here the derivative is a `derivWithin` over `Set.Ioi 0`
+since `β > 0`. -/
+theorem meanEnergy_eq_neg_deriv_log_mathZ_of_beta
```
Maybe make this one a lemma rather than a theorem?

or4nge19 committed:
fix comments

or4nge19 committed:
fix added space

or4nge19 committed:
rename CanonicalEnsemble field parameters following mathlib naming conventions

no other change

or4nge19 committed:
Revert "rename CanonicalEnsemble field parameters following mathlib naming conventions"

This reverts commit f3d9dfbc6cc856b41c073bb255b9f5aa6261e60f.

or4nge19 committed:
Reapply "rename CanonicalEnsemble field parameters following mathlib naming conventions"

This reverts commit 2e848ab742fc62ab8eb152405ae3baf675a5fbee.

or4nge19 committed:
Merge branch 'master' into or4nge19-CanonicalEnsemble

jstoobysmith commented on PhysLean/StatisticalMechanics/CanonicalEnsemble/Basic.lean:
```diff
@@ -10,35 +10,121 @@ import PhysLean.Meta.Informal.SemiFormal
+  phaseSpaceunit : ℝ := 1
+  /-- Assumption that the phase space unit is positive. -/
+  hPos : 0 < phaseSpaceunit := by positivity
+  energyMeasurable : Measurable energy
```
this is a lemma so should be energy_measurable

jstoobysmith reviewed (commented):
Add a single comment, otherwise all looks good

or4nge19 committed:
energyMeasurable -> energy_measurable + cleaning

jstoobysmith reviewed (approved):
Approved.
